server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -264,16 +271,42 @@ public void restoreSnapshot(final RestoreSnapshotRequest request,
                     metadataBuilder = Metadata.builder();
                 }
 
-                final List<IndexId> indexIdsInSnapshot = repositoryData.resolveIndices(indicesInSnapshot);
+                // set up feature state indices
+                Set<String> requestedFeatureStateIndexes = new HashSet<>();
+                Set<String> nonRequestedFeatureStateIndexes = new HashSet<>();
+
+                if (request.includeGlobalState() == false) {
+                    for (SnapshotFeatureInfo snapshotFeatureInfo : snapshotInfo.featureStates()) {
+                        if (Set.of(request.featureStates()).contains(snapshotFeatureInfo.getPluginName())) {
```

Comment:
I think it'll be more efficient to just do a `.contains()` on the list rather than converting to a `Set` since the lists of feature states are going to have a practical maximum of tens of elements.

(especially since the `Set.of` call is inside a loop, so it'll convert the list to a set N times where N=`snapshotInfo.featureStates().size()`)

Reply:
`request.featureStates()` returns a `String[]`, so there's no `contains` method. I'll keep the set but move its instantiation outside of the loop.